google/resumable_media/requests/download.py

```diff
@@ -286,3 +291,58 @@ def update(self, unused_chunk):
         Args:
             unused_chunk (bytes): A chunk of data.
         """
+
+
+def _add_decoder(response_raw, md5_hash):
+    """Patch the ``_decoder`` on a ``urllib3`` response.
+
+    This is so that we can intercept the compressed bytes before they are
+    decoded.
+
+    Only patches if the content encoding is ``gzip``.
+
+    Args:
+        response_raw (urllib3.response.HTTPResponse): The raw response for
+            an HTTP request.
+        md5_hash (Union[_DoNothingHash, hashlib.md5]): A hash function which
+            will get updated when it encounters compressed bytes.
+
+    Returns:
+        Union[_DoNothingHash, hashlib.md5]: Either the original ``md5_hash``
+        if ``_decoder`` is not patched. Otherwise, returns a ``_DoNothingHash``
+        since the caller will no longer need to hash to decoded bytes.
+    """
+    encoding = response_raw.headers.get(u'content-encoding', u'').lower()
+    if encoding != u'gzip':
```

Comment:
ISTR that the accept-encoding is set to "gzip, deflate".
I think if the server decides to send deflate content the code here won't work, and instead will need to handle both cases.

Reply:
I based this off the code [from `urllib3`][1]. I can just change it to

```python
if u'gzip' not in encoding:
    ...
```

However, `gzip, deflate` is a **list** of encodings intended for `Accept-Encoding`. This is the `Content-Encoding` of the bytes.

[1]: https://github.com/shazow/urllib3/blob/ff5c6749c237ff6a8b0a3450713ecce7469a7ea8/urllib3/response.py#L256-L258

Reply:
Right, but if the server sends data compressed using deflate encoding, the current code won't monkey patch the decoder to handle that case, which means we'll try to compare the checksum with the Requests-decoded data and get a checksum failure.

Reply:
This PR is **explicitly** about `gzip`. If you think we need to handle the `deflate` encoding, that's a separate discussion (and this is the first I'm hearing it).

Reply:
Yes, I didn't think of it before.

I believe everything will work right now because if you make a request to GCS with accept-encoding: 'gzip, deflate' GCS happens to send using gzip; but it would be spec-legal to send in deflate format instead, and that would cause the library to break in the way I noted above.

A simple soln would be to remove 'deflate' in the accept-encoding for now.
(Side note: I've read that deflate is somewhat more CPU efficient than gzip, which I think is more of a concern for browsers in general.)

Reply:
@mfschwartz Can you file an issue that we can follow up on? This change needs to go in soon so we can release a `0.3.1` and un-break at least **some subset** of people.